README.md
# ruler_picker_bn

Flutter package for ruler type value picker

## Screenshots

<img src="https://raw.githubusercontent.com/BoredNerdy/ruler_picker/main/screenshots/height.gif" height="400em" width="225em" />
<img src="https://raw.githubusercontent.com/BoredNerdy/ruler_picker/main/screenshots/width.gif" height="400em" width="225em" />

## Usage

[Example](https://github.com/BoredNerdy/ruler_picker/blob/main/example/lib/main.dart)

To use this package :

* add the dependency to your [pubspec.yaml](https://github.com/BoredNerdy/ruler_picker/blob/main/pubspec.yaml) file.

```yaml
  dependencies:
    flutter:
      sdk: flutter
    ruler_picker_bn:
```

### How to use

```dart
...
 SizedBox(
  width: 200,
  height: 75,
  child: RulerPicker(
                    onChange: (val) {
                        setState(() {
                          valueKG = val;
                        });
                    },
                    background: Colors.white,
                    lineColor: Colors.black,
                    direction: Axis.horizontal,
                    startValue: 70,
                    maxValue: 200,
                  ),
 ),
...
```

### List Of Properties
* background : set ruler background color
* lineColor  : set ruler value line color
* direction  : set ruler direction
* startValue : set ruler starting value position
* maxValue   : set ruler max value
* minValue   : set ruler min value
* padding    : set ruler padding value
* onChange   : event for get to ruler value


# License
Copyright (c) 2022 BoredNerdy

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


## Getting Started

For help getting started with Flutter, view our online [documentation](https://flutter.io/).

For help on editing package code, view the [documentation](https://flutter.io/developing-packages/).


CHANGELOG.md
## 0.0.1

* TODO: Describe initial release.

## 0.0.2

* readme file edited

## 0.0.3

* Made improvements for pub points.

## 0.0.4

* Made improvements for pub points.

## 0.0.5

* Example screenshots added.

## 0.0.6

* Made some performance improvements.


test/ruler_picker_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:ruler_picker_bn/ruler_picker_bn.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/ruler_picker.dart
library ruler_picker;

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:scrollable_positioned_list/scrollable_positioned_list.dart';

class RulerPicker extends StatefulWidget {
  const RulerPicker({Key? key,this.minValue=0,this.maxValue=250, this.startValue = 160, this.background = Colors.white, this.lineColor = Colors.black, this.padding = const EdgeInsets.all(4), this.direction = Axis.vertical, required this.onChange})
      : super(key: key);
  final ValueSetter<int> onChange;
  final Color background;
  final EdgeInsets padding;
  final Color lineColor;
  final Axis direction;
  final int startValue;
  final int minValue ;
  final int maxValue;
  @override
  RulerPickerState createState() => RulerPickerState();
}

class RulerPickerState extends State<RulerPicker> {
  final ItemScrollController controllerScroll = ItemScrollController();
  final ItemPositionsListener itemPositionsListener = ItemPositionsListener.create();

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(builder: (BuildContext context, BoxConstraints constraints) {
      return Stack(
        children: [
          Container(
            padding: widget.padding,
            decoration: BoxDecoration(color: widget.background),
            child: widget.direction == Axis.vertical
                ? Row(
                    crossAxisAlignment: CrossAxisAlignment.center,
                    children: [
                      Expanded(
                        child: ScrollablePositionedList.builder(
                            itemCount: 5000,
                            scrollDirection: widget.direction,
                            shrinkWrap: true,
                            initialScrollIndex: 2501,
                            itemScrollController: controllerScroll,
                            itemPositionsListener: itemPositionsListener,
                            itemBuilder: (context, index) {
                              return Row(
                                mainAxisAlignment: MainAxisAlignment.center,
                                mainAxisSize: MainAxisSize.min,
                                children: [
                                  Container(
                                    decoration: BoxDecoration(
                                      border: Border(
                                        top: widget.direction == Axis.vertical
                                            ? BorderSide(
                                                color: index % 10 == 0
                                                    ? widget.lineColor.withOpacity(0.54)
                                                    : index % 5 == 0
                                                        ? widget.lineColor.withOpacity(0.36)
                                                        : widget.lineColor.withOpacity(0.12))
                                            : BorderSide.none,
                                        left: widget.direction == Axis.horizontal
                                            ? BorderSide(
                                                color: index % 10 == 0
                                                    ? widget.lineColor.withOpacity(0.54)
                                                    : index % 5 == 0
                                                        ? widget.lineColor.withOpacity(0.36)
                                                        : widget.lineColor.withOpacity(0.12))
                                            : BorderSide.none,
                                      ),
                                    ),
                                    height: widget.direction == Axis.vertical
                                        ? 1
                                        : constraints.maxHeight *
                                                (index % 10 == 0
                                                    ? 1
                                                    : index % 5 == 0
                                                        ? 0.75
                                                        : 0.5) -
                                            16 -
                                            widget.padding.top -
                                            widget.padding.bottom,
                                    width: widget.direction == Axis.horizontal
                                        ? 1
                                        : constraints.maxWidth *
                                                (index % 10 == 0
                                                    ? 1
                                                    : index % 5 == 0
                                                        ? 0.75
                                                        : 0.5) -
                                            16 -
                                            widget.padding.left -
                                            widget.padding.right,
                                    margin: widget.direction == Axis.vertical ? const EdgeInsets.symmetric(vertical: 2) : const EdgeInsets.symmetric(horizontal: 2),
                                  ),
                                ],
                              );
                            }),
                      ),
                      Icon(
                        Icons.arrow_left,
                        color: Colors.black54,
                        size: 16,
                      ),
                    ],
                  )
                : Column(
                    crossAxisAlignment: CrossAxisAlignment.center,
                    children: [
                      Expanded(
                        child: ScrollablePositionedList.builder(
                            itemCount: 5000,
                            scrollDirection: widget.direction,
                            shrinkWrap: true,
                            initialScrollIndex: 2501,
                            itemScrollController: controllerScroll,
                            itemPositionsListener: itemPositionsListener,
                            itemBuilder: (context, index) {
                              return Row(
                                mainAxisAlignment: MainAxisAlignment.center,
                                mainAxisSize: MainAxisSize.min,
                                children: [
                                  Container(
                                    decoration: BoxDecoration(
                                      border: Border(
                                        top: widget.direction == Axis.vertical
                                            ? BorderSide(
                                                color: index % 10 == 0
                                                    ? widget.lineColor.withOpacity(0.54)
                                                    : index % 5 == 0
                                                        ? widget.lineColor.withOpacity(0.36)
                                                        : widget.lineColor.withOpacity(0.12))
                                            : BorderSide.none,
                                        left: widget.direction == Axis.horizontal
                                            ? BorderSide(
                                                color: index % 10 == 0
                                                    ? widget.lineColor.withOpacity(0.54)
                                                    : index % 5 == 0
                                                        ? widget.lineColor.withOpacity(0.36)
                                                        : widget.lineColor.withOpacity(0.12))
                                            : BorderSide.none,
                                      ),
                                    ),
                                    height: widget.direction == Axis.vertical
                                        ? 1
                                        : constraints.maxHeight *
                                                (index % 10 == 0
                                                    ? 1
                                                    : index % 5 == 0
                                                        ? 0.75
                                                        : 0.5) -
                                            16 -
                                            widget.padding.top -
                                            widget.padding.bottom,
                                    width: widget.direction == Axis.horizontal
                                        ? 1
                                        : constraints.maxWidth *
                                                (index % 10 == 0
                                                    ? 1
                                                    : index % 5 == 0
                                                        ? 0.75
                                                        : 0.5) -
                                            16 -
                                            widget.padding.left -
                                            widget.padding.right,
                                    margin: widget.direction == Axis.vertical ? const EdgeInsets.symmetric(vertical: 2) : const EdgeInsets.symmetric(horizontal: 2),
                                  ),
                                ],
                              );
                            }),
                      ),
                      Icon(
                        Icons.arrow_drop_up,
                        color: widget.lineColor.withOpacity(0.54),
                        size: 16,
                      ),
                    ],
                  ),
          ),
          Align(
            alignment: widget.direction == Axis.vertical ? Alignment.topCenter : Alignment.centerLeft,
            child: Container(
              width: widget.direction == Axis.vertical ? constraints.maxWidth : constraints.maxWidth * 0.2,
              height:widget.direction == Axis.vertical ? constraints.maxHeight * 0.2 : constraints.maxHeight,
              decoration: BoxDecoration(
                  gradient: LinearGradient(
                begin: widget.direction == Axis.vertical ? Alignment.topCenter : Alignment.centerLeft,
                end: widget.direction == Axis.vertical ? Alignment.bottomCenter : Alignment.centerRight,
                stops: [
                  0.2,
                  0.4,
                  0.6,
                  0.8,
                ],
                colors: [
                  widget.background,
                  widget.background.withOpacity(0.75),
                  widget.background.withOpacity(0.5),
                  widget.background.withOpacity(0.25),
                ],
              )),
            ),
          ),
          Align(
            alignment:widget.direction == Axis.vertical ? Alignment.bottomCenter : Alignment.centerRight,
            child: Container(

              width: widget.direction == Axis.vertical ? constraints.maxWidth : constraints.maxWidth * 0.2,
              height:widget.direction == Axis.vertical ? constraints.maxHeight * 0.2 : constraints.maxHeight,
              decoration: BoxDecoration(
                  gradient: LinearGradient(
                begin: widget.direction == Axis.vertical ? Alignment.topCenter : Alignment.centerLeft,
                end: widget.direction == Axis.vertical ? Alignment.bottomCenter : Alignment.centerRight,
                stops: [
                  0.2,
                  0.4,
                  0.6,
                  0.8,
                ],
                colors: [
                  widget.background.withOpacity(0.25),
                  widget.background.withOpacity(0.5),
                  widget.background.withOpacity(0.75),
                  widget.background,
                ],
              )),
            ),
          ),
        ],
      );
    });
  }

  @override
  void initState() {
    super.initState();
    itemPositionsListener.itemPositions.addListener(() {
      var positions = itemPositionsListener.itemPositions.value;
      int? min;
      int? max;
      if (positions.isNotEmpty) {
        min = positions
            .where((ItemPosition position) => position.itemTrailingEdge > 0)
            .reduce((ItemPosition min, ItemPosition position) => position.itemTrailingEdge < min.itemTrailingEdge ? position : min)
            .index;
        max = positions
            .where((ItemPosition position) => position.itemLeadingEdge < 1)
            .reduce((ItemPosition max, ItemPosition position) => position.itemLeadingEdge > max.itemLeadingEdge ? position : max)
            .index;
      }
      int position = ((((max! - min!) / 2)) + min).toInt() - 2520 + widget.startValue ;
      int value = position < widget.minValue
          ? widget.minValue
          : position > widget.maxValue
              ? widget.maxValue
              : position;
      widget.onChange(value);
    });
  }

  /* Widget get positionsView => ValueListenableBuilder<Iterable<ItemPosition>>(
        valueListenable: itemPositionsListener.itemPositions,
        builder: (context, positions, child) {
          int? min;
          int? max;
          if (positions.isNotEmpty) {
            min = positions
                .where((ItemPosition position) => position.itemTrailingEdge > 0)
                .reduce((ItemPosition min, ItemPosition position) => position.itemTrailingEdge < min.itemTrailingEdge ? position : min)
                .index;
            max = positions
                .where((ItemPosition position) => position.itemLeadingEdge < 1)
                .reduce((ItemPosition max, ItemPosition position) => position.itemLeadingEdge > max.itemLeadingEdge ? position : max)
                .index;
          }
          int position = ((((max! - min!) / 2)) + min).toInt() - 2400;
          int value = position < 0
              ? 0
              : position > 250
                  ? 250
                  : position;
            widget.returnValue = value;
          return Text(
            '${value.toString().padLeft(3,'0')} cm',
            style: TextStyle(fontSize: 12, fontWeight: FontWeight.bold, color: Colors.black87),
          );
        },
      );*/

}


examples/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.CreateAndShow(L"examples", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:ruler_picker_bn/ruler_picker_bn.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Ruler Picker Example',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Ruler Picker Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int valueKG = 0;
  int valueCM = 0;
  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.start,
              crossAxisAlignment: CrossAxisAlignment.center,
              mainAxisSize: MainAxisSize.min,
              children: [
                Padding(
                  padding: const EdgeInsets.only(right: 8.0),
                  child: Text('${valueKG.toString().padLeft(3, '0')} kg'),
                ),
                SizedBox(
                  width: 200,
                  height: 75,
                  child: RulerPicker(
                    onChange: (val) {
                      setState(() {
                        valueKG = val;
                      });
                    },
                    background: Colors.white,
                    lineColor: Colors.black,
                    direction: Axis.horizontal,
                    startValue: 70,
                    maxValue: 200,
                  ),
                ),
              ],
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.start,
              crossAxisAlignment: CrossAxisAlignment.center,
              mainAxisSize: MainAxisSize.min,
              children: [
                Padding(
                  padding: const EdgeInsets.only(right: 8.0),
                  child: Text('${valueCM.toString().padLeft(3, '0')} cm'),
                ),
                SizedBox(
                  width: 75,
                  height: 200,
                  child: RulerPicker(
                    onChange: (val) {
                      setState(() {
                        valueCM = val;
                      });
                    },
                    background: Colors.white,
                    lineColor: Colors.black,
                    direction: Axis.vertical,
                  ),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}
